### PR TITLE
Added public_ip method get_by_id and delete and unit tests

### PR DESCRIPTION
**Pull request by navalkp**
Created this PR as had some merge issues in old PR.

### CODE REVIEW

**Review by waynew — CHANGES_REQUESTED** [review]
**question** do we have integration tests for these?

File: src/saltext/vmware/modules/vmc_public_ip.py [hunk 1]
```diff
@@ -107,3 +107,138 @@ def get(
...
+    refresh_key,
+    authorization_host,
+    org_id,
+    sddc_id,
+    public_ip_id,
+    verify_ssl=True,
+    cert=None,
+):
+    """
+    Retrieves Public IPs from Given SDDC
```
Comment (waynew) [inline src/saltext/vmware/modules/vmc_public_ip.py:123]: Doc for this is identical to the `get` function (which should also be `list`). Either this one or `get` should be changed.

**Review by waynew — CHANGES_REQUESTED** [review]

**Comment by shaikres** [comment]
> **question** do we have integration tests for these?

yes in the next PR, i will add create, update functions and the integration tests for all functions of this module

**Comment by shaikres** [comment]
> I'm confused - what is the motivation to break up all of the testing for a single piece of functionality into two PRs?
> 
> Especially tests.
> 
> If the idea is just to provide a smaller PR, I appreciate that motivation, but it's a slight misunderstanding.
> 
> Tests are a form of documentation and explanation for a change. Not providing tests doesn't make the PR easier to understand - it makes it harder because there's less explanation about what the code should be doing.
> 
> If the goal is smaller, easier to review, a better approach is a WIP PR, providing a smaller chunk of functionality with complete tests. To use the crud example, one could provide a draft/ WIP PR and tests for:
> 
> 1. Create function - when that has been reviewed...
> 2. Delete function - when that has been reviewed...
> 3. Read (get/list functions)
> 4. Update function
> 
> For the PRs I've been seeing lately, I wouldn't say that's necessary, but if you wanted faster feedback that could be good.
> 
> --
> 
> Of course it's entirely possible I've completely misunderstood the desire behind not providing integration tests as well 😅

sure @waynew,
actually we have one WIP PR: [https://github.com/saltstack/salt-ext-modules-vmware/pull/226](url) which contains the remaining functions and complete integration tests for public IP module. We thought once this PR is merged, we will rebase the PR-226 and push for review.

should it be fine or shall i include those changes also in this PR itself?

**Comment by waynew** [comment]
**tl;dr** - integration tests for `get_by_id`† and delete should be added.

---

The *goal* is always clear communication.

When we open a PR we're attempting to communicate *what* code does, *why* it should be added to the project, and to future developers (even if that's just our own future selves) what we were thinking about when we initially wrote the code and why we made the decisions we made.

If we don't provide *meaningful* tests then we're not fully communicating.

In some cases, it's possible to write just unit tests and have them completely meaningful. As we've seen, most of our extension code is mostly just CRUD, and so a unit test is only part of the story. If we want to have meaningful and complete communication about our code, then integration tests help a *lot*.

I'm pretty sure that for our extension module we've already encountered some API changes (or maybe I'm just thinking of docs moving) that were uncovered by the integration tests.

If we want someone to really understand our code (and know what steps we've taken to test it) then we need to provide those actual tests. That gives them:

- information about what we were or were not thinking about when we wrote our code
- information about what in our code is important and essential and what is an implementation detail
- the ability to make changes in our code, like adding assertion statements, and _know_ that they've broken our expectations (or not)

Generally:  A PR for adding a function should have complete, **meaningful** tests that cover all of the branches for that particular function. In our case that *usually* means unit *and* integration tests.

---

† obviously it should be renamed to `get`, while `get` should be `list`, in keeping with typical Salt idioms.

**Comment by shaikres** [comment]
Added other functions and integration tests.

test checks are failing for Linux(3.7, 3.8 and 3.9) for the error: `ImportError: cannot import name 'Markup' from 'jinja2' (/home/runner/work/salt-ext-modules-vmware/salt-ext-modules-vmware/.nox/tests-3/lib/python3.9/site-packages/jinja2/__init__.py)`

**Review by waynew — CHANGES_REQUESTED** [review]
This looks mostly good - the requested changes are mostly just a find->replace, and super minor. I tested the module locally and can list/create/get/update/delete.

- [ ] - rename `public_ip_id` to `id` throughout
- [ ] - rename `public_ip_name` to `name` throughout
- [ ] - move id/name parameters to the start of the functions
- [ ] -  update tests as necessary
- [ ] - add period to the end of sentences in docstrings
- [ ] - fix `ip` - `IP` casing in docstrings

I *think* that's a comprehensive list of the suggestions. As soon as these changes are made I can re-run the full test suite and merge :+1: 

I think that's

File: src/saltext/vmware/modules/vmc_public_ip.py [hunk 2]
```diff
@@ -102,8 +103,291 @@ def get(
...
     )
+
+
+def get(
+    hostname,
+    refresh_key,
+    authorization_host,
+    org_id,
+    sddc_id,
+    public_ip_id,
```
[the remark this answers is not in the document]
Comment (shaikres) [reply src/saltext/vmware/modules/vmc_public_ip.py:119]: updated

File: src/saltext/vmware/modules/vmc_public_ip.py [hunk 3]
```diff
@@ -102,8 +103,291 @@ def get(
...
+    verify_ssl=True,
+    cert=None,
+):
+    """
+    Delete public ip from Given SDDC
```
[the remark this answers is not in the document]
Comment (shaikres) [reply src/saltext/vmware/modules/vmc_public_ip.py:191]: updated

File: src/saltext/vmware/modules/vmc_public_ip.py [hunk 4]
```diff
@@ -102,8 +103,291 @@ def get(
...
+        hostname to get access token
+
+    org_id
+        org_id of the SDDC
+
+    sddc_id
+        sddc_id from which public ip will be deleted
+
+    public_ip_id
+        Id of specific public IP to be deleted
```
[the remark this answers is not in the document]
Comment (shaikres) [reply src/saltext/vmware/modules/vmc_public_ip.py:215]: updated

File: src/saltext/vmware/modules/vmc_public_ip.py [hunk 5]
```diff
@@ -102,8 +103,291 @@ def get(
...
+
```
[the remark this answers is not in the document]
Comment (shaikres) [reply src/saltext/vmware/modules/vmc_public_ip.py:216]: added period

File: src/saltext/vmware/modules/vmc_public_ip.py [hunk 6]
```diff
@@ -102,8 +103,291 @@ def get(
...
+    refresh_key,
+    authorization_host,
+    org_id,
+    sddc_id,
+    public_ip_name,
+    verify_ssl=True,
+    cert=None,
+):
+    """
+    Create public ip for Given SDDC
```
[the remark this answers is not in the document]
Comment (shaikres) [reply src/saltext/vmware/modules/vmc_public_ip.py:259]: updated

File: src/saltext/vmware/modules/vmc_public_ip.py [hunk 7]
```diff
@@ -102,8 +103,291 @@ def get(
...
+    )
+
+
+def create(
+    hostname,
+    refresh_key,
+    authorization_host,
+    org_id,
+    sddc_id,
+    public_ip_name,
```
[the remark this answers is not in the document]
Comment (shaikres) [reply src/saltext/vmware/modules/vmc_public_ip.py:254]: updated

File: src/saltext/vmware/modules/vmc_public_ip.py [hunk 8]
```diff
@@ -102,8 +103,291 @@ def get(
...
+        hostname to get access token
+
+    org_id
+        org_id of the SDDC
+
+    sddc_id
+        sddc_id for which public ip should be retrieved
+
+    public_ip_id
+        public ip id
```
[the remark this answers is not in the document]
Comment (shaikres) [reply src/saltext/vmware/modules/vmc_public_ip.py:356]: updated

File: src/saltext/vmware/modules/vmc_public_ip.py [hunk 9]
```diff
@@ -102,8 +103,291 @@ def get(
...
+
+    public_ip_name
+        name of public ip it will be updated
```
[the remark this answers is not in the document]
Comment (shaikres) [reply src/saltext/vmware/modules/vmc_public_ip.py:359]: updated

File: src/saltext/vmware/modules/vmc_public_ip.py [hunk 10]
```diff
@@ -102,8 +103,291 @@ def get(
...
+        hostname to get access token
+
+    org_id
+        org_id of the SDDC
+
+    sddc_id
+        sddc_id from which public ip should be retrieved
+
+    public_ip_id
+        public ip id for which details should be retrieved
```
[the remark this answers is not in the document]
Comment (shaikres) [reply src/saltext/vmware/modules/vmc_public_ip.py:148]: updated

**Comment by shaikres** [comment]
> This looks mostly good - the requested changes are mostly just a find->replace, and super minor. I tested the module locally and can list/create/get/update/delete.
> 
> * [ ]  - rename `public_ip_id` to `id` throughout
> * [ ]  - rename `public_ip_name` to `name` throughout
> * [ ]  - move id/name parameters to the start of the functions
> * [ ]  -  update tests as necessary
> * [ ]  - add period to the end of sentences in docstrings
> * [ ]  - fix `ip` - `IP` casing in docstrings
> 
> I _think_ that's a comprehensive list of the suggestions. As soon as these changes are made I can re-run the full test suite and merge 👍
> 
> I think that's

Updated the code as per all the suggestions.

**Comment by waynew** [comment]
The jinja failures are fixed on the salt master branch. I installed that version of salt locally and ran the full test suite so this is :+1:

**Review by waynew — APPROVED** [review]
